pkg/datapath/loader/compile.go

```diff
@@ -363,8 +363,8 @@ func compileDatapath(ctx context.Context, dirs *directoryInfo, isHost bool, logg
 			if ctx.Err() == nil {
 				scopedLog.WithField(logfields.Params, logfields.Repr(p)).
 					WithError(err).Debug("JoinEP: Failed to compile")
+				return err
 			}
-			return err
```

Comment:
^^ I just noticed that the review was draft (because that PR itself is still draft), so I posted the comment just now.

Reply:
@joestringer The good news is context cancellation now propagates to the loader code as intended, so we should be able to bail out properly now. However, the changes to `compileDatapath()` do cause trouble down the line:

```
cilium-combined-logs.txt:2022-07-06T14:28:19.381956708Z level=error msg="Command execution failed" cmd="[llc -mattr=dwarfris -march=bpf -mcpu=v2 -filetype=obj -o /var/run/cilium/state/templates/f2716579d3201fb10b2c21d3cd45087ab44160473fa1f3178ae515214c30fdca/bpf_lxc.o]" error="signal: killed" subsys=datapath-loader
cilium-combined-logs.txt:2022-07-06T14:28:19.381958982Z level=error msg="Failed to compile bpf_lxc.o: Command execution failed for [llc -mattr=dwarfris -march=bpf -mcpu=v2 -filetype=obj -o /var/run/cilium/state/templates/f2716579d3201fb10b2c21d3cd45087ab44160473fa1f3178ae515214c30fdca/bpf_lxc.o]: context canceled" compiler-pid=621 linker-pid=622 subsys=datapath-loader
cilium-combined-logs.txt:2022-07-06T14:28:19.381991491Z level=error msg="Error while rewriting endpoint BPF program" containerID= datapathPolicyRevision=0 desiredPolicyRevision=1 endpointID=3847 error="failed to parse ELF file /var/run/cilium/state/templates/f2716579d3201fb10b2c21d3cd45087ab44160473fa1f3178ae515214c30fdca/bpf_lxc.o: unable to open ELF: EOF" identity=4 ipv4= ipv6= k8sPodName=/ subsys=endpoint
```

The last line `failed to parse ELF file ...` is repeated hundreds of times because the pipeline now thinks compilation succeeded when it was actually cancelled. (this happens rather often in CI VMs under load)

Attached one such failure here: [0632a6d8_K8sKafkaPolicyTest_Kafka_Policy_Tests_KafkaPolicies.zip](https://github.com/cilium/cilium/files/9064184/0632a6d8_K8sKafkaPolicyTest_Kafka_Policy_Tests_KafkaPolicies.zip)

Failures during Kafka deploy/image pull seem _really_ common, this leads me to think they might be load-induced.

Reply:
@ti-mo There was a reason I made those changes around returning errors  - https://github.com/cilium/cilium/pull/18267#issuecomment-996254405. The agent code path at many places isn't checking for cancelled context, and that's why they log the error messages even when the failures are expected . :(